chromium_src/chrome/browser/profiles/delete_profile_helper.cc

```diff
@@ -0,0 +1,46 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/sync/service/brave_sync_service_impl.h"
+#include "chrome/browser/sync/sync_service_factory.h"
+#include "components/sync/service/sync_user_settings.h"
+
+// Include to prevent redefining HasPrimaryAccount method at the header
+#include "components/signin/public/identity_manager/identity_manager.h"
+
+class Profile;
+
+namespace {
+
+bool StopSyncIfActive(Profile* profile) {
+  if (SyncServiceFactory::HasSyncService(profile)) {
+    syncer::BraveSyncServiceImpl* sync_service =
+        static_cast<syncer::BraveSyncServiceImpl*>(
+            SyncServiceFactory::GetForProfile(profile));
+    sync_service->StopAndClear();
+  }
+
+  return true;
+}
+
+}  // namespace
+
+// Suppress warning after macro expanding
+// error: '&&' within '||' [-Werror,-Wlogical-op-parentheses]
+// if (!identity_manager || !identity_manager->HasPrimaryAccount(..) &&
+//     !StopSyncIfActive(profile)) {
+// Compiler want us to make
+// if (!A || (!B && C)), but with this type of override we can't.
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wlogical-op-parentheses"
+
+#define HasPrimaryAccount(LEVEL) \
```

Comment:
Currently it is only `DisableSyncForProfileDeletion` calling `HasPrimaryAccount` and if future changes adding more to call `HasPrimaryAccount`, wouldn't we do `StopAndClear` excessively?

Can we observe `PrimaryAccountManager::Observer::OnPrimaryAccountChanged` to see if event source is `signin_metrics::ProfileSignout` and then do `StopAndClear`? `identity_manager->GetPrimaryAccountMutator()->ClearPrimaryAccount()` will eventually fire `OnPrimaryAccountChanged`

Reply:
>Currently it is only DisableSyncForProfileDeletion calling HasPrimaryAccount and if future changes adding more to call HasPrimaryAccount, wouldn't we do StopAndClear excessively?

Basically yes, if in the future other places will at the `src/chrome/browser/profiles/delete_profile_helper.cc` will call `HasPrimaryAccount` there is a chansthat we will call StopAndClear excessively. This will be the thing we don't want.
```#define HasPrimaryAccount(LEVEL) \
  HasPrimaryAccount(LEVEL) && StopSyncIfActive(profile)
```
uses `profile` variable, so go get the failure of extra `HasPrimaryAccount` in the future, we must have `profile` near that place also. `profile` variable is in many places `delete_profile_helper.cc` by the obvious reasons.

I had idea to change `HasPrimaryAccount` override to capture `identity_manager` variable, but `HasPrimaryAccount` will be invoked through  `identity_manager->HasPrimaryAccount()`, so no much sense here.

I tried to avoid use of patches, plus recent comment https://github.com/brave/brave-core/pull/24446/files#r1661628792 , so I did similar here.

It seems the only robust way is to have the patch.

> Can we observe PrimaryAccountManager::Observer::OnPrimaryAccountChanged to see if event source is signin_metrics::ProfileSignout and then do StopAndClear? identity_manager->GetPrimaryAccountMutator()->ClearPrimaryAccount() will eventually fire OnPrimaryAccountChanged

Infortunatelly, this approach doesn't work.
The problem is at 
```
void DisableSyncForProfileDeletion(Profile* profile) {
  signin::IdentityManager* identity_manager =
      IdentityManagerFactory::GetForProfileIfExists(profile);
  if (!identity_manager ||
      !identity_manager->HasPrimaryAccount(signin::ConsentLevel::kSignin)) {
    return;
  }
```
`HasPrimaryAccount` returns false and we do early exit out of function.

In addition we override [BravePrimaryAccountMutatorImpl::ClearPrimaryAccount](https://github.com/brave/brave-core/blob/master/components/signin/internal/identity_manager/brave_primary_account_mutator_impl.cc#L31) with an empty implementation.

Reply:
@darkdh I have tried this https://github.com/brave/brave-core/pull/24988/commits/9f569c6defcd85fe4a810151f1c79d104f65b7c1

basically at `StopSyncIfActive` we check for __func__ from macro, it should be `DisableSyncForProfileDeletion` plus the additional `static_assert` around `IdentityManager* identity_manager`  to fail the compilation if it is outside of the `DisableSyncForProfileDeletion`.

wdyt, does it look ok?

Reply:
That looks better and `StopSyncIfActive` should be enough, static_assert around IdentityManager can be false positive easily.